### DISCUSSION TITLE
How to deal with distorted tracker bar spacings?

### DISCUSSION THREAD

**Discussion by pfefferniels** [pianoroll/roll-image-parser#3] [General]
This is an extract from a scan of one of the rolls I received from Marc Widuch, marked up using the markholes program.

<img width="1326" alt="grafik" src="https://github.com/user-attachments/assets/60e8096f-5462-4edc-b916-9d6e2bbb8457">

As you can see, the expression holes on the left and right sides, for whatever reason, do not align with the tracker bar lines: instead of being on the line, they are in between. The musical notes themselves, however, match perfectly with the tracker lines. This leads to incorrect track numbers being assigned.

Does anyone have an idea why this is happening (is it somethign common? Perhaps it's happening due to non-straight paper during the scanning process?) and how to best address it?

(`analyzeTrackerBarSpacing()` calculated the `holeSeparation` to 19.179, which is half of what the actual spacing is. For the image above I forced it manually to 38.359 ...)

**Comment by craigsapp** [comment]
I have seen this very occasionally with the red Welte rolls (I think on the order of 1%), although different scanning/brands may have different probabilities).

It sounds like a failure in the Fourier analysis within `analyzeTrackerBarSpacing()`, where a sub-harmonic at 19.179 is higher in the spectrum than the true spacing approximately twice as wide.  The above markup on the tracker bar lines is for 38.359 (why 9 instead of 8)?  If so, then this value is slightly too small, causing it to be off by 1/2 of a hole eventually.

> Perhaps it's happening due to non-straight paper during the scanning process?

Plotting the absolute value of the Fourier analysis might graphically demonstrate what the problem is.  The roll/scan is very straight, so I do not think it is a problem with the image.   There is a large gap between notes on the right side of the roll and the expression holes which might be a factor (but I don't think so).

It would be good to plot the Fourier analysis to see what the problem is (where the problem is most likely happening).

> The musical notes themselves, however, match perfectly with the tracker lines. 

They do not match perfectly:  I calculate the spacing with a Fourier transform, but there was a problem using the phase of the transform to assign the absolute position on the paper, so I used linear regression (or similar) to find the best position for the tracker bar hole alignments.   This will tend to favor the notes region, centering the incorrect spacings on the notes and showing errors in the expression track on the edges.   Here is the tracker bar position for the right-most (highest?) note:

<img width="146" alt="Screenshot 2024-12-02 at 05 31 09" src="https://github.com/user-attachments/assets/9a20b4cb-af8c-47ac-972b-f794ad78dfb8">

Compare to about 30% from the left, where the center of the tracker bars was probably assigned:

<img width="90" alt="Screenshot 2024-12-02 at 05 32 53" src="https://github.com/user-attachments/assets/38c23e11-2715-4713-9931-550bf59996a4">

The top image shows the tracker line favoring the left side of the holes, while the bottom one is closer to the center (not exactly, so exactly at the offset position measured after determining the spacing from Fourier analysis.

--------------------------

There are two functions that are run before `analyzeTrackerBarSpacing()`:

https://github.com/pianoroll/roll-image-parser/blob/1bd87eff61b070a68fc4f578a3a8d1fc086cd6cd/src/RollImage.cpp#L179-L187

So there are possible problems in the previously called functions `storeCorrectedCentroidHistogram()` and `analyzeRawRowPositions()`, but I don't think as likely as in `analyzeTrackerBarSpacing()`.  There are some functions after calling `analyszeTrackerBarSpacing()` which are related to trying to fix problems in the analysis, I think.

If there is no code for detecting 1/2 subharmonic yet, then it could be added: if there are significant amounts of detected holes that are halfway between detected tracker lines, then the subharmonic was likely chosen, and the Fourier transform peak should be checked for at twice the detected frequency.
